File: benchmarking/tests/passport_processing_02/solution/solution.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define KEYS_LEN 7
const char* const keys[] = { "byr:", "iyr:", "eyr:", "hgt:", "hcl:", "ecl:", "pid:" };

bool is_valid_field(const char *key, const char *value)
{
    if (strncmp(key, "byr", 3) == 0) {
        int v = atoi(value);
        return v >= 1920 && v <= 2002;
    }

    if (strncmp(key, "iyr", 3) == 0) {
        int v = atoi(value);
        return v >= 2010 && v <= 2020;
    }

    if (strncmp(key, "eyr", 3) == 0) {
        int v = atoi(value);
        return v >= 2020 && v <= 2030;
    }

    if (strncmp(key, "hgt", 3) == 0) {
        int v = atoi(value);
        if (strncmp(value + 3, "cm", 2) == 0)
        {
            return v >= 150 && v <= 193;
        }

        if (strncmp(value + 2, "in", 2) == 0)
        {
            return v >= 59 && v <= 76;
        }

        return false;
    }

    if (strncmp(key, "hcl", 3) == 0) {
        if (value[0] != '#') {
            return false;
        }

        for (int i = 1; i < 7; i++) {
            if (!(value[i] >= 48 && value[i] <= 57) && !(value[i] >= 97 && value[i] <= 102)) {
                return false;
            }
        }

        return true;
    }

    if (strncmp(key, "ecl", 3) == 0) {
        return strncmp(value, "amb", 3) == 0
            || strncmp(value, "blu", 3) == 0
            || strncmp(value, "brn", 3) == 0
            || strncmp(value, "gry", 3) == 0
            || strncmp(value, "grn", 3) == 0
            || strncmp(value, "hzl", 3) == 0
            || strncmp(value, "oth", 3) == 0;
    }

    if (strncmp(key, "pid", 3) == 0) {
        for (int i = 0; i < 9; i++) {
            if (!(value[i] >= 48 && value[i] <= 57)) {
                return false;
            }
        }

        return !(value[9] >= 48 && value[9] <= 57);
    }

    return false;
}

bool is_valid_passport(const char *pass)
{
    for (int i = 0; i < KEYS_LEN; i++) {
        const char *p = strstr(pass, keys[i]);
        if (!p) {
            return false;
        }

        if (!is_valid_field(p, p + 4)) {
            return false;
        }
    }

    return true;
}

int count_valid_passports(const char *filename)
{
    FILE *file = fopen(filename, "r");

    // Include space for newline and string terminator
    char buffer[128] = { 0 };

    uint8_t values = 0;
    int correct = 0;
    while (fgets(buffer, 128, file)) {
        if (buffer[0] == '\n') {
            if (values == 0x7f) {
                correct++;
            }

            values = 0;
        }

        for (int i = 0; i < KEYS_LEN; i++) {
            const char *p = strstr(buffer, keys[i]);
            if (!p) {
                continue;
            }

            if (is_valid_field(p, p + 4)) {
                values = values | (0x1 << i);
            }
        }
    }

    if (values == 0x7f) {
        correct++;
    }

    fclose(file);

    return correct;
}

File: benchmarking/tests/passport_processing_02/solution/solution.py
import re

# define the set of mandatory fields
mandatory_fields = {
    'byr': [(re.compile('^([0-9]{4})$'), 1920, 2002)],
    'iyr': [(re.compile('^([0-9]{4})$'), 2010, 2020)],
    'eyr': [(re.compile('^([0-9]{4})$'), 2020, 2030)],
    'hgt': [(re.compile('^([0-9]+)cm$'), 150, 193), (re.compile('^([0-9]+)in$'), 59, 78)],
    'hcl': [(re.compile('^#[0-9a-f]{6}$'), None, None)],
    'ecl': [(re.compile('^amb|blu|brn|gry|grn|hzl|oth$'), None, None)],
    'pid': [(re.compile('^[0-9]{9}$'), None, None)],
}

def is_valid_field(field, value):

    """
    >>> is_valid_field('byr', '2002')
    True
    >>> is_valid_field('byr', '2003')
    False

    >>> is_valid_field('hgt', '60in')
    True
    >>> is_valid_field('hgt', '190cm')
    True
    >>> is_valid_field('hgt', '190in')
    False
    >>> is_valid_field('hgt', '190')
    False

    >>> is_valid_field('hcl', '#123abc')
    True
    >>> is_valid_field('hcl', '#123abz')
    False
    >>> is_valid_field('hcl', '123abc')
    False

    >>> is_valid_field('ecl', 'brn')
    True
    >>> is_valid_field('ecl', 'wat')
    False

    >>> is_valid_field('pid', '000000001')
    True
    >>> is_valid_field('pid', '0123456789')
    False
    """

    if field not in mandatory_fields:
        return False

    for regex, lower, upper in mandatory_fields[field]:
        result = regex.match(value)
        if result:

            # match without number is correct
            groups = result.groups()
            if not groups:
                return True

            # check if number is between limits
            value = groups[0]
            if value.isdigit() and lower <= int(value) <= upper:
                return True

    # no match with any of the rules for the field
    return False

def is_valid_passport(passport):

    """
    >>> is_valid_passport('eyr:1972 cid:100 hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926')
    False
    >>> is_valid_passport('iyr:2019 hcl:#602927 eyr:1967 hgt:170cm ecl:grn pid:012533040 byr:1946')
    False
    >>> is_valid_passport('hcl:dab227 iyr:2012 ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277')
    False
    >>> is_valid_passport('hgt:59cm ecl:zzz eyr:2038 hcl:74454a iyr:2023 pid:3556412378 byr:2007')
    False

    >>> is_valid_passport('pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980 hcl:#623a2f')
    True
    >>> is_valid_passport('eyr:2029 ecl:blu cid:129 byr:1989 iyr:2014 pid:896056539 hcl:#a97842 hgt:165cm')
    True
    >>> is_valid_passport('hcl:#888785 hgt:164cm byr:2001 iyr:2015 cid:88 pid:545766238 ecl:hzl eyr:2022')
    True
    >>> is_valid_passport('iyr:2010 hgt:158cm hcl:#b6652a ecl:blu byr:1944 eyr:2021 pid:093154719')
    True
    """

    # extract the set of fields from the passport
    fields = {field.split(':')[0]:field.split(':')[1] for field in passport.split()}

    # check if all mandatory fields are present
    if not set(mandatory_fields) <= set(fields):
        return False

    # check if all mandatory fields are valid
    return all(
        is_valid_field(field, value)
        for field, value in fields.items() if field in mandatory_fields
    )

def passports(filename):

    """
    >>> list(passports('invalid_passports.txt'))
    ['eyr:1972 cid:100 hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926', 'iyr:2019 hcl:#602927 eyr:1967 hgt:170cm ecl:grn pid:012533040 byr:1946', 'hcl:dab227 iyr:2012 ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277', 'hgt:59cm ecl:zzz eyr:2038 hcl:74454a iyr:2023 pid:3556412378 byr:2007']
    """

    with open(filename) as lines:
        passport = []
        for line in lines:
            line = line.rstrip('\n')
            if line:
                passport.append(line)
            elif passport:
                yield ' '.join(passport)
                passport = []

        if passport:
            yield ' '.join(passport)

def count_valid_passports(filename):

    """
    >>> count_valid_passports('invalid_passports.txt')
    0
    >>> count_valid_passports('valid_passports.txt')
    4
    >>> count_valid_passports('adventofcode.input.txt')
    145
    """

    # count the number of valid passports in the given file
    return sum(is_valid_passport(passport) for passport in passports(filename))

if __name__ == '__main__':
    import doctest
    doctest.testmod()
